test/common/config/udpa_resource_test.cc

```diff
@@ -55,30 +62,135 @@ TEST(UdpaResourceNameTest, DecodeSuccess) {
   EXPECT_EQ("%#[]&=", resource_name.context().params().at("foo"));
 }
 
-// Validate that the URI decoding behaves with a near-empty UDPA resource name.
-TEST(UdpaResourceNameTest, DecodeEmpty) {
+// Validate that URL decoding behaves as expected component-wise.
+TEST(UdpaResourceLocatorTest, DecodeSuccess) {
+  const auto resource_locator =
+      UdpaResourceIdentifier::decodeUrl(EscapedUrlWithManyQueryParamsAndDirectives);
+  EXPECT_EQ("f123%/?#o", resource_locator.authority());
+  EXPECT_EQ("envoy.config.listener.v3.Listener", resource_locator.resource_type());
+  EXPECT_EQ(3, resource_locator.id().size());
```

Comment:
Done, but not for directives, since when working with proto types you don't get for free equality and nice debug string output. It's possible to do this with some extra work, but not worth it in this case IMHO.